_posts/2019-06-25-dynamic_linear_models_tfprobability/dynamic_linear_models_tfprobability.Rmd

```diff
@@ -0,0 +1,805 @@
+
+```{r}
+optimizer <- tf$compat$v1$train$AdamOptimizer(0.1)
+
+# only train on the training set!    
+loss_and_dists <- ts_train %>% sts_build_factored_variational_loss(model = model)
+variational_loss <- loss_and_dists[[1]]
+
+train_op <- optimizer$minimize(variational_loss)
+```
+
+Note how the loss is defined on the training set only, not the complete series.
+
+Now to actually train the model, we create a session and run that operation:
+
+```{r}
+ with (tf$Session() %as% sess,  {
+      
+      sess$run(tf$compat$v1$global_variables_initializer())
+   
+      for (step in 1:n_iterations) {
+        res <- sess$run(train_op)
+        loss <- sess$run(variational_loss)
+        if (step %% 10 == 0)
+          cat("Loss: ", as.numeric(loss), "\n")
+      }
+ })
+```
+
+
+Given we have that session, let's make use of it and compute all the estimates we desire.
+Again, -- the following snippets will end up in the `fit_with_vi` function, so don't run them in isolation just yet.
+
+#### Obtaining forecasts
+
+The first thing we want for the model to give us are forecasts. In order to create them, it needs _samples from the posterior_. Luckily we already have the posterior distributions, returned from `sts_build_factored_variational_loss`, so let's sample from them and pass them to [sts_forecast](https://rstudio.github.io/tfprobability/reference/sts_forecast.html):
+
+```{r}
+variational_distributions <- loss_and_dists[[2]]
+posterior_samples <-
+  Map(
+    function(d) d %>% tfd_sample(n_param_samples),
+    variational_distributions %>% reticulate::py_to_r() %>% unname()
+  )
+forecast_dists <- ts_train %>% sts_forecast(model, posterior_samples, n_forecast_steps)
+``` 
+
+`sts_forecast()` returns distributions, so we call `tfd_mean()` to get the posterior predictions and `tfd_stddev()` for the corresponding standard deviations:
+
+```{r}
+fc_means <- forecast_dists %>% tfd_mean()
+fc_sds <- forecast_dists %>% tfd_stddev()
+``` 
+
+By the way -- as we have the full posterior distributions, we are by no means restricted to summary statistics! We could easily use `tfd_sample()` to obtain individual forecasts.
+
+#### Smoothing and filtering (Kálmán filter)
+
+Now, the second (and last, for this example) thing we will want are the smoothed and filtered regression coefficients. The famous [Kálmán Filter](https://en.wikipedia.org/wiki/Kalman_filter) is a Bayesian-in-spirit method where at each time step, predictions are corrected by how much they differ from an incoming observation. _Filtering_ estimates are based on observations we've seen so far; _smoothing estimates_ are computed "in hindsight", making use of the complete time series.
```

Comment:
>  _Filtering_ estimates are based on observations we've seen so far; _smoothing estimates_ are computed "in hindsight", making use of the complete time series.

Either italicize "estimates" with both "Filtering" and "smoothing," or with neither.